heron/downloaders/src/java/com/twitter/heron/downloader/Extractor.java

```diff
@@ -26,9 +26,15 @@
 import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
 import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream;
 
-final class Extractor {
+class Extractor {
 
-  static void extract(InputStream in, Path destination) throws IOException {
+  private static final Extractor INSTANCE = new Extractor();
```

Comment:
What's the reason to use singleton pattern here? In general, I think we should avoid using this patter. It causes a lot of pain to maintain code related with the singleton.

Reply:
I changed it from static method to a class method of a singleton for unit testing. so I can easily mock Extractor for unit testing.

Reply:
If it's just for easier unit testing purpose, you don't need to change the original static method.

PowerMockito could help you mock static method easily. Here's an example: https://github.com/twitter/heron/blob/master/heron/scheduler-core/tests/java/com/twitter/heron/scheduler/client/SchedulerClientFactoryTest.java#L82

And you can search 'PowerMockito.' for more examples. I suggest we move away from the singleton pattern here and use powermockito for testing.